nifi-nar-bundles/nifi-spark-bundle/nifi-livy-processors/src/main/java/org/apache/nifi/processors/livy/ExecuteSparkInteractive.java

```diff
@@ -178,48 +247,188 @@ public void onTrigger(ProcessContext context, final ProcessSession session) thro
+
                 try {
-                    final JSONObject output = result.getJSONObject("data");
-                    flowFile = session.write(flowFile, out -> out.write(output.toString().getBytes()));
-                    flowFile = session.putAttribute(flowFile, CoreAttributes.MIME_TYPE.key(), LivySessionService.APPLICATION_JSON);
-                    session.transfer(flowFile, REL_SUCCESS);
-                } catch (JSONException je) {
-                    // The result doesn't contain the data, just send the output object as the flow file content to failure (after penalizing)
-                    log.error("Spark Session returned an error, sending the output JSON object as the flow file content to failure (after penalizing)");
-                    flowFile = session.write(flowFile, out -> out.write(result.toString().getBytes()));
+
+                    final JSONObject jsonResponseObj = new JSONObject(jsonResponse);
+
+                    Map<String, String> headers = new HashMap<>();
+                    headers.put("Content-Type", LivySessionService.APPLICATION_JSON);
+                    headers.put("X-Requested-By", LivySessionService.USER);
+                    headers.put("Accept", "application/json");
+
+                    JSONObject jobInfo = readJSONObjectFromUrl(jsonResponseObj.getString("url"), livySessionService, headers);
+
+                    flowFile = session.write(flowFile, out -> out.write(jobInfo.toString().getBytes()));
                     flowFile = session.putAttribute(flowFile, CoreAttributes.MIME_TYPE.key(), LivySessionService.APPLICATION_JSON);
-                    flowFile = session.penalize(flowFile);
+
+                    Thread.sleep(statusCheckInterval);
+
+                    String state  = jobInfo.getString("state");
+                    log.debug(" ====> jsonResponseObj State: " + state);
+
+                    switch (state) {
+                        case "success":
+                            log.debug(" ====> success State: " + state);
+                            session.transfer(flowFile, REL_SUCCESS);
+                            break;
+                        case "dead":
+                            log.debug(" ====> dead State: " + state);
+                            session.transfer(flowFile, REL_FAILURE);
+                            break;
+                        default:
+                            log.debug(" ====> default State: " + state);
+                            session.transfer(flowFile, REL_WAIT);
+                            break;
+                    }
+
+                } catch (JSONException | InterruptedException e) {
+
+                    //Incoming flow file is not an JSON file hence consider it to be an triggering point
+
+                    String batchPayload = "{ \"pyFiles\": [\"" +context.getProperty(PY_FILES).getValue()+ "\"], " +
+                            "\"file\" : \""+context.getProperty(MAIN_PY_FILE).getValue()+"\" }";
```

Comment:
This is confusing to me. Why are we saying that if the incoming flowfile is not a valid JSON, we are going ahead with the assumption that it is going to be PySpark? I mean the assumption here lacks clarity. Please correct me, if I'm wrong.

Reply:
Could you please check the description @ https://issues.apache.org/jira/browse/NIFI-4946

The assumption was made such that it doesn't break existing code flow and at the same time we wanted to know the status of the submitted job.

So the naive idea was to re-route the Livy Json response back to Spark processor only, so that it can get last submitted `url` from the custom (tampered) JSON response, wait for user specified wait time and again query the Livy for the Job status in a loop till it succeeds or fails.

So when the processor is configured to submit a Spark job, it will expect the incoming flowfile to be an custom Json response with an `url` field to query the Livy, if not it is considered as a triggering point nothing else.

I am open for any ideas from your end.

Thanks.

Reply:
@zenfenan could please review above logic and suggest a way to handle plain Scala/Python code and packages source files for pyfiles/jars ?

Reply:
The code can be submitted using the `CODE` property, right? That used to work. Or are you asking of a way to upload/send files or jars through Livy?

Reply:
As per the code flow @ https://issues.apache.org/jira/browse/NIFI-4946, currently I am able to send *.zip files (Python modules) through livy. My question was what should we do with flowfile, when we are using the processor to submit a batch job?

Reply:
Sometime this week I am planning to add support for jar files, args and application name over the Livy options.

The catch here is unlike plain Spark code, batch process code will take time to finish which is expected one as we know. So as a hack I was re-routing the Json response after batch submission to itself, where I poll the incoming flowfile and check whether it is a Json file and if so I will try to get the "livy url" to post again to know the status of the batch job as long as it runs. After knowing the the job finished, the success route is triggered.

That was the reason the I have made an assumption if the incoming file is Json, it is from previous batch job submission.

In short, the flow file :
- Is considered as a triggering point (or)
- Is considered as plain Spark code that compiles over Livy (or)
- Is a Livy Json response, which can further be used to check the status of long running Spark batch job

I was looking for the right Nifi way of handling this? 

I feel I am too conservative and trying to fit all the functionalities on one processor.
- Flow file/property can be used to run a Spark code
- Pyfiles can be used to run Spark batch job
- Jars can be used to run Spark batch job
- Args options for batch mode
- By rerouting the success to itself, it can monitor the long running job over Livy rest APIs

Reply:
Since the current processor is called ExecuteSparkInteractive, you could move the batch functionality out into something called SubmitSparkJob or something. The outgoing flow file could contain the application ID and/or any other information that would allow you to monitor the job downstream (perhaps with InvokeHttp through Livy, e.g.) Are you still interested in working on this? It seems like very nice functionality to have in NiFi!